Document compiling V8 from source (#459)
hashintel/hash: packages/engine/README.md

export RUST_LOG=debug
```

> If your simulation takes a lot of memory and is using JavaScript behaviors, the JavaScript runner might run out of memory.

Reviewer: ```suggestion
If your simulation requires a lot of memory and uses JavaScript behaviors, the JavaScript runner may run out of memory.
```

> The first step you can take is passing a larger heap size to the runner:

Reviewer: ```suggestion
As a first step, you can provide a larger heap size to the runner:
```

```shell

> cargo run --bin cli -- --project /path/to/my-hash-project single-run --num-steps <NUM-STEPS> --js-runner-max-heap-size <NEW-SIZE-IN-MB>

Reviewer: Let's not specify the type of experiment we want to run (`single-run`)
```suggestion
cargo run --bin cli -- --js-runner-max-heap-size <NEW-SIZE-IN-MB> <YOUR-CLI-ARGUMENTS>
```

```

> This will increase the heap size, but you may still run into limitations.

Reviewer 2: ```suggestion
This will increase the heap size, but you may still run into limitations beyond 4GB.
```

> This will increase the heap size but you may still run into limitations.

Reviewer: ```suggestion
This will increase the heap size, but you may still run into limitations.
```

> The next step is to recompile the underlying JavaScript engine and setting flags for it:

Reviewer: ```suggestion
The next step is to recompile V8, the underlying JavaScript engine, and set flags for it:
```

```shell
export V8_FROM_SOURCE = "1"
export GN_ARGS = "v8_enable_pointer_compression=false v8_enable_shared_ro_heap=true"
```

- `V8_FROM_SOURCE` will force the V8 engine to be compiled from source and not use a pre-compiled version.

> This will take a fairly long time, expect 15min minimum. This can be mitigated in the subsequent compilations by using [sccache](https://github.com/mozilla/sccache) or [ccache](https://ccache.dev/). Our build scripts will detect and use them. Set the `$SCCACHE` or `$CCACHE` environmental variable if it's not in your path.

Reviewer: ```suggestion
This will take quite a long time, expect at least 15 minutes. This can be mitigated in subsequent compiles by using [sccache](https://github.com/mozilla/sccache) or [ccache](https://ccache.dev/). Our build scripts will detect and use them. Set the environment variable `$SCCACHE` or `$CCACHE` if the binary is not in your `$PATH`.
```

> - `v8_enable_pointer_compression` is an optimization reducing RAM usage at the cost of max heap size.

Reviewer: ```suggestion
- `v8_enable_pointer_compression` is an optimization that reduces RAM usage at the expense of maximum heap size.
```

> - `v8_enable_pointer_compression` is an optimization reducing RAM usage at the cost of max heap size.

Reviewer: What does pointer compression specifically?

Author: V8 is 64bits so all pointers are 8 bytes long.
Pointer compression will cut these 8 bytes in half. The higher 4 bytes are stored in a register and are common to all pointers.
Then each pointer uses 4 bytes for the "partial address" and the remaining 4 are used to store values like small integers or tags.
When a pointer is used V8 will add the 4 higher bytes from the register to the "partial address" and get a "real memory address".

Reviewer: Can you add a description to the readme?

Author: Is it useful? This is going very deep into how V8 works.

Reviewer: I don't think we need the whole description, but we want more than just "reduces RAM, increases max heap size". Something like

```
- `v8_enable_pointer_compression` is an optimization reducing RAM usage at the cost of max heap size.
  Disabling this will allow addressing more than 4 gigabytes of RAM.
```

Reviewer: I think that's fine, thank you!

> - `v8_enable_shared_ro_heap` is required to disable pointer compression and allows the read-only memory to be shared across V8 isolate.

Reviewer: ```suggestion
- `v8_enable_shared_ro_heap` is required to disable pointer compression and enable read-only memory sharing by V8 isolates. This means, that read-only memory may be shared across different workers for JavaScript.
```

Reviewer: We are creating an `Isolate` at each worker. There is one JS Runner per worker, each JS Runner runs `run_experiment` (_runner/javascript/mod.rs_, not _experiment/controller/run.rs_). I only clarified based on your description.

Author: I was wondering if creating a new platform would prevent this sharing between `Isolate`. But I guess not.

Reviewer: What do you mean by "platform"?

Author: ```rust
let platform = v8::new_default_platform(0, false).make_shared();
v8::V8::initialize_platform(platform);
```

Reviewer: And I guess the platform is not shared across runners?
If this is an optimization we currently can't use we shouldn't mention the flag in the readme.

Author: I'm not actually sure if the platform is shared or not. At first I thought it wasn't but there might be only one and we initialize it if it's not already done.

We have to mention the flags, not because of its own value but because you can't disable pointer compression without it. V8 will not compile.

Reviewer: > At first I thought it wasn't but there might be only one and we initialize it if it's not already done.

Could you figure this out?

> We have to mention the flag, not because of its own value but because you can't disable pointer compression without it. V8 will not compile.

This should be mentioned IMO.

Author: > Could you figure this out?

I've looked into it before replying earlier, let's just say the documentation is pretty sparse.
It seems to handle allocation, clock, async task.
So not a supervisor to `Isolate` as I first thought, more something unrelated.

Reviewer: Verified in source: 

https://github.com/v8/v8/blob/eb9a19a0a11d7767888d632897f348ff36873841/src/heap/read-only-heap.cc#L68-L110

To summarize: With pointer compression, there is one `ReadOnlyHeap` per `Isolate`. Without PC, there is only one `ReadOnlyHeap` shared between all `Isolates`. There have a global `Mutex` and are completely independent from `Platform`, even though, `Platform` manages jobs.

> Mentioned V8 [5e5b0f7](https://github.com/hashintel/hash/commit/5e5b0f702018c5782f9a1c0d584d33087e0975ef)

I would put it the other way around: First describe, what this flag does, then say, that this is required if pointer compression is disabled. No strong opinion on that, though.

Author: Swapped the order https://github.com/hashintel/hash/pull/459/commits/1a4b19f24e996d0a65bae3c38ea89dff1c315edd

> - `v8_enable_shared_ro_heap` enables read-only memory sharing by V8 isolates. This means, that read-only memory may be shared across different workers for JavaScript. It is required to compile V8 without pointer compression.

Reviewer: ```suggestion
- `v8_enable_shared_ro_heap` enables read-only memory sharing by V8 isolates. This means, that read-only memory may be shared across different workers for JavaScript. Enabling this is required to compile V8 without pointer compression.
```

> - `v8_enable_shared_ro_heap` enables read-only memory sharing by V8 isolates. This means, that read-only memory may be shared across different workers for JavaScript. Enabling this is required to compile V8 without pointer compression.

Reviewer 2: I genuinely don't know (hence me asking 😄) is this actually true? Do we have only one V8 runtime and start an isolate per worker, or do we start a whole V8 thing per worker. It definitely feels like we should be starting isolates if that's not what we're doing

Author: We are starting isolates, the rest seems to be globals that are only initialized the first time.

Reviewer 2: Good to know :)
